hedera-node/src/main/java/com/hedera/services/txns/util/RandomGenerateTransitionLogic.java

```diff
@@ -0,0 +1,115 @@
+import com.hedera.services.context.TransactionContext;
+import com.hedera.services.context.properties.GlobalDynamicProperties;
+import com.hedera.services.stream.RecordsRunningHashLeaf;
+import com.hedera.services.txns.TransitionLogic;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.nio.ByteBuffer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_RANDOM_GENERATE_RANGE;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.SUCCESS;
+
+/**
+ * Provides the state transition for generating a pseudorandom bytes or pseudorandom number.
+ */
+@Singleton
+public class RandomGenerateTransitionLogic implements TransitionLogic {
+	private static final Logger log = LogManager.getLogger(RandomGenerateTransitionLogic.class);
+
+	private final TransactionContext txnCtx;
+	private final SideEffectsTracker sideEffectsTracker;
+	private final Supplier<RecordsRunningHashLeaf> runningHashLeafSupplier;
+
+	private final GlobalDynamicProperties properties;
+
+	@Inject
+	public RandomGenerateTransitionLogic(final TransactionContext txnCtx,
+			final SideEffectsTracker sideEffectsTracker,
+			final Supplier<RecordsRunningHashLeaf> runningHashLeafSupplier,
+			final GlobalDynamicProperties properties) {
+		this.txnCtx = txnCtx;
+		this.sideEffectsTracker = sideEffectsTracker;
+		this.runningHashLeafSupplier = runningHashLeafSupplier;
+		this.properties = properties;
+	}
+
+	@Override
+	public void doStateTransition() {
+		if (!properties.isRandomGenerationEnabled()) {
+			return;
+		}
+		final var op = txnCtx.accessor().getTxn().getRandomGenerate();
+
+		// Use n-3 running hash instead of n-1 running hash for processing transactions quickly
+		final var nMinus3RunningHash = runningHashLeafSupplier.get().getNMinus3RunningHash();
+		if (nMinus3RunningHash == null || nMinus3RunningHash.getHash() == null) {
+			log.info("No n-3 record running hash available to generate random number");
+			return;
+		}
+		//generate binary string from the running hash of records
+		final var pseudoRandomBytes = nMinus3RunningHash.getHash().getValue();
+
+		final var range = op.getRange();
+		if (range > 0) {
+			// generate pseudorandom number in the given range
+			final var initialBitsValue = Math.abs(ByteBuffer.wrap(pseudoRandomBytes, 0, 4).getInt());
+			int pseudoRandomNumber = (int) ((range * (long) initialBitsValue) >>> 32);
+
+			sideEffectsTracker.trackRandomNumber(pseudoRandomNumber);
+		} else {
+			sideEffectsTracker.trackRandomBytes(pseudoRandomBytes);
+		}
+
+		txnCtx.setStatus(SUCCESS);
```

Comment:
Since you've added `RandomGenerate` [here](https://github.com/hashgraph/hedera-services/blob/b425fb5e5508a8dd0efc68b15063d176a243d8ce/hedera-node/src/main/java/com/hedera/services/txns/TransitionRunner.java#L84), this isn't necessary.

Reply:
Ah yes ..removed